Added the ability to specify a proxy in the config.json file

TheraNinjaCat opened the pull request:
All that needs to be done to enable a proxy is to specify a key `"proxy"` in one of the config.json files, and set it to a format accepted by electrons default [`setProxy`](https://www.electronjs.org/docs/api/session#sessetproxyconfig) function.
i.e. `"proxy": "socks5://localhost:8080"`

Fixes https://github.com/vector-im/element-desktop/issues/735

<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## ✨ Features
 * Added the ability to specify a proxy in the config.json file ([\#58](https://github.com/vector-im/element-desktop/pull/58)). Fixes #735. Contributed by @TheraNinjaCat.<!-- CHANGELOG_PREVIEW_END -->

t3chguy commented:
Needs DCO sign-off https://github.com/matrix-org/matrix-js-sdk/blob/master/CONTRIBUTING.rst#sign-off

TheraNinjaCat commented:
I assume a comment here works for that too.
Signed-off-by: Greg Best <greg@thebest.nz>

t3chguy commented:
Yup, that is fine :)

jryans commented on src/electron-main.js:
```diff
@@ -696,7 +696,17 @@ app.on('ready', async () => {
+
+    if (vectorConfig['proxy']) {
+      console.log(`Starting electron through proxy: ${vectorConfig['proxy']}`);
+      mainWindow.webContents.session.setProxy({proxyRules: vectorConfig['proxy']})
```
You should be able to use `await` here, and then you can remove the `then` handler and put `loadURL` back where it was outside of the `if` to cover both cases.

NHAS commented:
Has there been any more discussion about adding proper proxy support to Element Desktop?

We've been using the `--proxy-server=address:port` command line option that Electron supports by default, and I imagine that many people who don't have a direct connection (e.g ISP/Work Place level proxy) will be doing the same. 

Unfortunately the auto update function does not respect this command line flag and will attempt to connect directly, potentially revealing that the device is using Element/Electron which impacts privacy. 
So people who use this flag (as they may have to) might be unknowingly putting themselves at risk. 

From a functionality perspective, this also seems rather important to have. This would both enable people who are behind proxies by default to access matrix, and enable greater privacy by allowing less technical people to use proxies.

kn0wmad commented:
> Has there been any more discussion about adding proper proxy support to Element Desktop?
> 
> We've been using the `--proxy-server=address:port` command line option that Electron supports by default, and I imagine that many people who don't have a direct connection (e.g ISP/Work Place level proxy) will be doing the same.
> 
> Unfortunately the auto update function does not respect this command line flag and will attempt to connect directly, potentially revealing that the device is using Element/Electron which impacts privacy.
> So people who use this flag (as they may have to) might be unknowingly putting themselves at risk.
> 
> From a functionality perspective, this also seems rather important to have. This would both enable people who are behind proxies by default to access matrix, and enable greater privacy by allowing less technical people to use proxies.

Bump.  Our users have to use the same workaround and get the same issues.

TheraNinjaCat commented:
I’ve not done any changes to this one, it’s been a while and I’ve not kept up to date with the codebase.
The change in the original PR here doesn’t actually fix the auto update proxy leak. As far as I could tell that’s an issue with Electron’s default auto-updater not respecting Electron’s proxy settings. I did look into proposing the replacement that’s suggested in the Electron documentation (that in my testing seemed to fix the issue) but it involved changing more upstream stuff around how updates are handled, however I didn’t feel like a change like that would be received well so I didn’t continue further.

TheraNinjaCat commented:
I have revisited this, moved the proxy handling to a helper module, and added basic support for the `file:` URI scheme.

I have moved the config from `config.json` to `electron-config.json` because the former seemed to be more relevant to the vector webapp running inside electron, but proxy settings are more relevant to electron itself.

I have also investigated the autoUpdater as this did not seem to obey the proxy settings from my original implementation, nor the `--proxy-server=address:port` chromium flag. It turns out the built-in autoUpdater does not support proxying at all, however most electron documentation suggests [using the `electron-updater` instead](https://www.electron.build/auto-update.html), which exposes the `netSession` object which can have proxy settings applied. As far as I can tell these two sessions are the only sessions in the app, however if it turns out that there are more it's trivial to add them to the helper to apply proxy settings to them as well.

The change to `electron-updater` does potentially break the current build/deploy/auto-update functionality as it no longer uses the same method of checking for updates. Where electron autoUpdater relies on a server responding with specific HTTP status codes, the `electron-updater` simply deploys a `latest-platform.yml` file for each platform that contains information for updating.

In regards to a test framework, I don't think the lack of a test suite should hold back implementing this, as due to several comments here people are already using the built-in chromium flag for using a proxy (which coincidentally uses the exact underlying APIs used by my changes), using a proxy isn't necessarily for the intention of increasing privacy as inside a walled-garden network it's often required for access to the internet, and if the underlying APIs stop working in the future, this will be due to an upstream change in electron or chromium.

In terms of writing tests for this, it would be possible to perform tests by using [`session.resolveProxy(url)`](https://www.electronjs.org/docs/api/session#sesresolveproxyurl), and validating that the output matches the defined proxy configuration.

Finally, I have also documented the proxy config inside the README.md file.

TheraNinjaCat commented:
I have just noticed that a bunch of packages were updated and I didn't notice before committing, including electron core, which I believe also deprecated enableRemoteModule, hence it being removed. All of those changes don't belong in this PR so after feedback has come back I can revert those changes.

xanoni commented:
Can confirm that the package auto updater just doesn't want to use the proxy, so I had to firewall everything that wasn't going to localhost:8118.

I tried using `torsocks`, `proxychains`, and I exported the `ALL_PROXY` and `HTTPS_PROXY` environment variables .... that all didn't help.

For all other Element traffic, the `--proxy-server=127.0.0.1:8118` parameter does the job.

NHAS commented:
Yup of course. You guys manage a massive amount of comms, so I appreciate that you definitely don't have time to re-read every each time you come back to something. 

I'll see if I can bug the maintainer of this issue to add it to the issue description, if that would be sufficient?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/element-hq/element-desktop?pullRequest=58) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/element-hq/element-desktop?pullRequest=58) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/element-hq/element-desktop?pullRequest=58) it.</sub>

dbkr commented:
This has been sitting since 2021 so I think it's time to close it. This isn't saying, 'no' to the feature, but this PR has 3 years worth of conflicts so it's probably as easy to start afresh anyway.

If anyone picks this up, some points from when we were discussing it:
 * An obvious use of this is to put traffic through Tor or similar, so we need to
   1. Make sure it doesn't break (ie. find a way to test it, and test it well)
   2. Make sure *all* the traffic the app sends uses the proxy. ie. this PR covers the web traffic + auto updater. Is that everything? Do we make any requests from the main process itself? Will we in the future?
 * The config passed to this is quite large. It's also just taking the proxy config object and exposing it in the config. Is this interface stable?
 * I forget the details of electron auto updating, but having to import a package we previously never pulled in seems very surprising. This would need a lot of explanation in comments as Travis was getting at. What's lacking about the one from `electron`?
 * The PAC file watching is a little 😳. This would definitely need tests. It also feels weird that it's our job.

That said, a good argument for this is that, if, in practice, people are using --proxy-server and leaking update traffic over a clearnet then this will help.
